
**<h1 align="center">Mining App Reviews</h1>**

---
I have included images of sample outputs where applicable to increase the readability of my Colab notebook.

You can collapse all sections by pressing ' Ctrl+] ' or by going in the Colab toolbar to View -> Collapse Sections.

#### **GATHERING REVIEWS**

##### Imports:

In [ ]:
!pip install -qq google-play-scraper
from google_play_scraper import app
import pandas as pd
import numpy as np
from google_play_scraper import Sort, reviews_all, reviews

##### My App - Parlera

In [ ]:
p_reviews = reviews_all(
    'com.enjoyingfoss.parlera',
    sleep_milliseconds=0,
    lang='en',
    country='us',
    sort=Sort.NEWEST,
)
df_parlera = pd.DataFrame(np.array(p_reviews),columns=['review'])
df_parlera.insert(0, 'Package Name', 'com.enjoyingfoss.parlera')
df_parlera = df_parlera.join(pd.DataFrame(df_parlera.pop('review').tolist()))
df_parlera.drop('reviewId', inplace=True, axis=1)
df_parlera.drop('userImage', inplace=True, axis=1)
df_parlera.drop('thumbsUpCount', inplace=True, axis=1)
df_parlera.drop('reviewCreatedVersion', inplace=True, axis=1)
df_parlera.drop('at', inplace=True, axis=1)
df_parlera.drop('replyContent', inplace=True, axis=1)
df_parlera.drop('repliedAt', inplace=True, axis=1)
dict = {'userName': 'Reviewer Name',
        'content': 'Review',
        'score': 'Rating'}
df_parlera.rename(columns=dict, inplace=True)
df_parlera.to_csv('com.enjoyingfoss.parlera.csv', index=None, header=True)

##### Similar App 1 - Heads Up!

In [ ]:
sa1_reviews = []
h_reviews, _ = reviews(
    'com.wb.headsup',
    lang='en',
    country='us',
    sort=Sort.NEWEST,
    count=5000
)
sa1_reviews.extend(h_reviews)
df_headsup = pd.DataFrame(sa1_reviews)
df_headsup.insert(0, 'Package Name', 'com.wb.headsup')
df_headsup.drop('reviewId', inplace=True, axis=1)
df_headsup.drop('userImage', inplace=True, axis=1)
df_headsup.drop('thumbsUpCount', inplace=True, axis=1)
df_headsup.drop('reviewCreatedVersion', inplace=True, axis=1)
df_headsup.drop('at', inplace=True, axis=1)
df_headsup.drop('replyContent', inplace=True, axis=1)
df_headsup.drop('repliedAt', inplace=True, axis=1)
dict = {'userName': 'Reviewer Name',
        'content': 'Review',
        'score': 'Rating'}
df_headsup.rename(columns=dict, inplace=True)
df_headsup.to_csv('com.wb.headsup.csv', index=None, header=True)

##### Similar App 2 - Wordscapes

In [ ]:
sa2_reviews = []
w_reviews, _ = reviews(
    'com.peoplefun.wordcross',
    lang='en',
    country='us',
    sort=Sort.NEWEST,
    count=5000
)
sa2_reviews.extend(w_reviews)
df_wordscapes = pd.DataFrame(sa2_reviews)
df_wordscapes.insert(0, 'Package Name', 'com.peoplefun.wordcross')
df_wordscapes.drop('reviewId', inplace=True, axis=1)
df_wordscapes.drop('userImage', inplace=True, axis=1)
df_wordscapes.drop('thumbsUpCount', inplace=True, axis=1)
df_wordscapes.drop('reviewCreatedVersion', inplace=True, axis=1)
df_wordscapes.drop('at', inplace=True, axis=1)
df_wordscapes.drop('replyContent', inplace=True, axis=1)
df_wordscapes.drop('repliedAt', inplace=True, axis=1)
dict = {'userName': 'Reviewer Name',
        'content': 'Review',
        'score': 'Rating'}
df_wordscapes.rename(columns=dict, inplace=True)
df_wordscapes.to_csv('com.peoplefun.wordcross.csv', index=None, header=True)

##### Competitor App 1 - Party Games for Groups

In [ ]:
g_reviews = reviews_all(
    'dh3games.partygamesforgroups',
    sleep_milliseconds=0,
    lang='en',
    country='us',
    sort=Sort.NEWEST,
)
df_partygamesforgroups = pd.DataFrame(np.array(g_reviews),columns=['review'])
df_partygamesforgroups.insert(0, 'Package Name', 'dh3games.partygamesforgroups')
df_partygamesforgroups = df_partygamesforgroups.join(pd.DataFrame(df_partygamesforgroups.pop('review').tolist()))
df_partygamesforgroups.drop('reviewId', inplace=True, axis=1)
df_partygamesforgroups.drop('userImage', inplace=True, axis=1)
df_partygamesforgroups.drop('thumbsUpCount', inplace=True, axis=1)
df_partygamesforgroups.drop('reviewCreatedVersion', inplace=True, axis=1)
df_partygamesforgroups.drop('at', inplace=True, axis=1)
df_partygamesforgroups.drop('replyContent', inplace=True, axis=1)
df_partygamesforgroups.drop('repliedAt', inplace=True, axis=1)
dict = {'userName': 'Reviewer Name',
        'content': 'Review',
        'score': 'Rating'}
df_partygamesforgroups.rename(columns=dict, inplace=True)
df_partygamesforgroups.to_csv('dh3games.partygamesforgroups.csv', index=None, header=True)

##### Competitor App 2 - Quinb

In [ ]:
q_reviews = reviews_all(
    'xyz.deepdaikon.quinb',
    sleep_milliseconds=0,
    lang='en',
    country='us',
    sort=Sort.NEWEST,
)
df_quinb = pd.DataFrame(np.array(q_reviews),columns=['review'])
df_quinb.insert(0, 'Package Name', 'xyz.deepdaikon.quinb')
df_quinb = df_quinb.join(pd.DataFrame(df_quinb.pop('review').tolist()))
df_quinb.drop('reviewId', inplace=True, axis=1)
df_quinb.drop('userImage', inplace=True, axis=1)
df_quinb.drop('thumbsUpCount', inplace=True, axis=1)
df_quinb.drop('reviewCreatedVersion', inplace=True, axis=1)
df_quinb.drop('at', inplace=True, axis=1)
df_quinb.drop('replyContent', inplace=True, axis=1)
df_quinb.drop('repliedAt', inplace=True, axis=1)
dict = {'userName': 'Reviewer Name',
        'content': 'Review',
        'score': 'Rating'}
df_quinb.rename(columns=dict, inplace=True)
df_quinb.to_csv('xyz.deepdaikon.quinb.csv', index=None, header=True)

##### Sample .csv File Output:

See [README](https://github.com/josephspagnuolo/App-Reviews?tab=readme-ov-file#sample-csv-file-output) file.

#### **PREPROCESSING THE TEXT**

##### Imports and Defined Functions:

In [ ]:
!pip install -qq contractions
!pip install -qq nltk
!pip install -qq num2words
import nltk
import string
import contractions
from num2words import num2words
from nltk.tokenize import wordpunct_tokenize
from nltk.corpus import stopwords, wordnet
from nltk.stem import WordNetLemmatizer
import re
import unicodedata
nltk.download('stopwords')
nltk.download('averaged_perceptron_tagger')
nltk.download('wordnet')
stop_words = set(stopwords.words('english'))
stop_words.add('game')
wnl = WordNetLemmatizer()

def sep_num_words_func(text):
    return re.sub(r"([0-9]+(\.[0-9]+)?)",r" \1 ", text).strip()

def remove_accented_chars_func(text):
    return unicodedata.normalize('NFKD', text).encode('ascii', 'ignore').decode('utf-8', 'ignore')

def remove_extra_whitespaces_func(text): 
    return re.sub(r'^\s*|\s\s*', ' ', text).strip()

def num_to_words(text):
    after_spliting = text.split()
    for index in range(len(after_spliting)):
        if after_spliting[index].isdigit():
            after_spliting[index] = num2words(after_spliting[index])
    numbers_to_words = ' '.join(after_spliting)
    return numbers_to_words

def get_wordnet_pos(tag):
    if tag.startswith('J'):
        return wordnet.ADJ
    elif tag.startswith('V'):
        return wordnet.VERB
    elif tag.startswith('N'):
        return wordnet.NOUN
    elif tag.startswith('R'):
        return wordnet.ADV
    else:
        return wordnet.NOUN


##### My App - Parlera

In [ ]:
df1 = pd.read_csv('com.enjoyingfoss.parlera.csv')

# I. Remove punctuations
# II. Remove special characters and emojis

df1['Contractions'] = df1['Review'].apply(lambda x: [contractions.fix(word) for word in x.split()])
df1['Back to String'] = [' '.join(map(str, l)) for l in df1['Contractions']]
df1['Back to String'].replace('1st', 'first', regex=True, inplace=True)
df1['Back to String'].replace('2nd', 'second', regex=True, inplace=True)
df1['Back to String'].replace('3rd', 'third', regex=True, inplace=True)
df1['Separated'] = df1['Back to String'].apply(sep_num_words_func)
df1['Punctuation'] = df1['Separated'].str.replace(r'[^\w\s]+', "  ")
df1['Punctuation'] = df1['Punctuation'].apply(remove_accented_chars_func)

# III. Turn numbers into text

df1['Convert Numbers to Words'] = df1['Punctuation'].apply(num_to_words)
df1['Convert Numbers to Words'] = df1['Convert Numbers to Words'].str.replace(r'[^\w\s]+', "  ")

# IV. Remove extra white spaces

df1['Convert Numbers to Words'] = df1['Convert Numbers to Words'].apply(remove_extra_whitespaces_func)

# V. Turn all words into lowercase

df1['Lowercase'] = df1['Convert Numbers to Words'].str.lower()

# VI. Remove stop words

df1['Tokenized'] = df1['Lowercase'].apply(wordpunct_tokenize)
df1['Stopwords'] = df1['Tokenized'].apply(lambda x: [word for word in x if word not in stop_words])

# VII. Lemmatize the reviews

df1['POS Tags'] = df1['Stopwords'].apply(nltk.tag.pos_tag)
df1['Wordnet'] = df1['POS Tags'].apply(lambda x: [(word, get_wordnet_pos(pos_tag)) for (word, pos_tag) in x])
df1['Final Pre-processed Review'] = df1['Wordnet'].apply(lambda x: [wnl.lemmatize(word, tag) for word, tag in x])
df1['Final Pre-processed Review'] = df1['Final Pre-processed Review'].apply(lambda x: [word for word in x if word not in stop_words])

# VIII. Output 15 sample pre-processed reviews

df1.drop('Reviewer Name', inplace=True, axis=1)
df1.drop('Rating', inplace=True, axis=1)
dict2 = {'Review': 'Original Review'}
df1.rename(columns=dict2, inplace=True)
df1.drop('Contractions', inplace=True, axis=1)
df1.drop('Back to String', inplace=True, axis=1)
df1.drop('Separated', inplace=True, axis=1)
df1.drop('Punctuation', inplace=True, axis=1)
df1.drop('Convert Numbers to Words', inplace=True, axis=1)
df1.drop('Lowercase', inplace=True, axis=1)
df1.drop('Tokenized', inplace=True, axis=1)
df1.drop('Stopwords', inplace=True, axis=1)
df1.drop('POS Tags', inplace=True, axis=1)
df1.drop('Wordnet', inplace=True, axis=1)
df1.head(15)

##### Similar App 1 - Heads Up!

In [ ]:
df2 = pd.read_csv('com.wb.headsup.csv')

# I. Remove punctuations
# II. Remove special characters and emojis

df2['Contractions'] = df2['Review'].apply(lambda x: [contractions.fix(word) for word in x.split()])
df2['Back to String'] = [' '.join(map(str, l)) for l in df2['Contractions']]
df2['Back to String'].replace('1st', 'first', regex=True, inplace=True)
df2['Back to String'].replace('2nd', 'second', regex=True, inplace=True)
df2['Back to String'].replace('3rd', 'third', regex=True, inplace=True)
df2['Separated'] = df2['Back to String'].apply(sep_num_words_func)
df2['Punctuation'] = df2['Separated'].str.replace(r'[^\w\s]+', "  ")
df2['Punctuation'] = df2['Punctuation'].apply(remove_accented_chars_func)

# III. Turn numbers into text

df2['Convert Numbers to Words'] = df2['Punctuation'].apply(num_to_words)
df2['Convert Numbers to Words'] = df2['Convert Numbers to Words'].str.replace(r'[^\w\s]+', "  ")

# IV. Remove extra white spaces

df2['Convert Numbers to Words'] = df2['Convert Numbers to Words'].apply(remove_extra_whitespaces_func)

# V. Turn all words into lowercase

df2['Lowercase'] = df2['Convert Numbers to Words'].str.lower()

# VI. Remove stop words

df2['Tokenized'] = df2['Lowercase'].apply(wordpunct_tokenize)
df2['Stopwords'] = df2['Tokenized'].apply(lambda x: [word for word in x if word not in stop_words])

# VII. Lemmatize the reviews

df2['POS Tags'] = df2['Stopwords'].apply(nltk.tag.pos_tag)
df2['Wordnet'] = df2['POS Tags'].apply(lambda x: [(word, get_wordnet_pos(pos_tag)) for (word, pos_tag) in x])
df2['Final Pre-processed Review'] = df2['Wordnet'].apply(lambda x: [wnl.lemmatize(word, tag) for word, tag in x])
df2['Final Pre-processed Review'] = df2['Final Pre-processed Review'].apply(lambda x: [word for word in x if word not in stop_words])

# VIII. Output 15 sample pre-processed reviews

df2.drop('Reviewer Name', inplace=True, axis=1)
df2.drop('Rating', inplace=True, axis=1)
dict2 = {'Review': 'Original Review'}
df2.rename(columns=dict2, inplace=True)
df2.drop('Contractions', inplace=True, axis=1)
df2.drop('Back to String', inplace=True, axis=1)
df2.drop('Separated', inplace=True, axis=1)
df2.drop('Punctuation', inplace=True, axis=1)
df2.drop('Convert Numbers to Words', inplace=True, axis=1)
df2.drop('Lowercase', inplace=True, axis=1)
df2.drop('Tokenized', inplace=True, axis=1)
df2.drop('Stopwords', inplace=True, axis=1)
df2.drop('POS Tags', inplace=True, axis=1)
df2.drop('Wordnet', inplace=True, axis=1)
df2.head(15)

<ipython-input-9-b5c233300750>:12: FutureWarning: The default value of regex will change from True to False in a future version.
  df2['Punctuation'] = df2['Separated'].str.replace(r'[^\w\s]+', "  ")
<ipython-input-9-b5c233300750>:18: FutureWarning: The default value of regex will change from True to False in a future version.
  df2['Convert Numbers to Words'] = df2['Convert Numbers to Words'].str.replace(r'[^\w\s]+', "  ")


Package Name                                    Original Review  \
0   com.wb.headsup         It wouldn't Load and it kept getting stuck   
1   com.wb.headsup  This game is taking my money this very bad. An...   
2   com.wb.headsup                 It has such a hard time opening...   
3   com.wb.headsup  This app sucks, won't open, downloads about 90...   
4   com.wb.headsup                              does not lode hate it   
5   com.wb.headsup                  Somethings are premium so its a 3   
6   com.wb.headsup  This game does not work for my phone don't rec...   
7   com.wb.headsup            App hangs on the splash screen. Period.   
8   com.wb.headsup                            Amazing would recommend   
9   com.wb.headsup  I love this game to play with my friends but w...   
10  com.wb.headsup  Doesn't even let you play the game it just mak...   
11  com.wb.headsup                                              Great   
12  com.wb.headsup                                 Wont load probably   
13  com.wb.headsup                                       It is good 👍   
14  com.wb.headsup  Change menu back. The old menu with multiple d...   

                           Final Pre-processed Review  
0                     [would, load, kept, get, stuck]  
1   [take, money, bad, everyone, money, police, sh...  
2                                  [hard, time, open]  
3   [app, suck, open, downloads, ninety, stop, poi...  
4                                        [lode, hate]  
5                         [something, premium, three]  
6                            [work, phone, recommend]  
7                 [app, hang, splash, screen, period]  
8                           [amaze, would, recommend]  
9   [love, play, friend, updated, deck, go, cannot...  
10          [even, let, play, make, want, pay, money]  
11                                            [great]  
12                                   [load, probably]  
13                                             [good]  
14  [change, menu, back, old, menu, multiple, deck...

##### Similar App 2 - Wordscapes

In [ ]:
df3 = pd.read_csv('com.peoplefun.wordcross.csv')

# I. Remove punctuations
# II. Remove special characters and emojis

df3['Contractions'] = df3['Review'].apply(lambda x: [contractions.fix(word) for word in x.split()])
df3['Back to String'] = [' '.join(map(str, l)) for l in df3['Contractions']]
df3['Back to String'].replace('1st', 'first', regex=True, inplace=True)
df3['Back to String'].replace('2nd', 'second', regex=True, inplace=True)
df3['Back to String'].replace('3rd', 'third', regex=True, inplace=True)
df3['Separated'] = df3['Back to String'].apply(sep_num_words_func)
df3['Punctuation'] = df3['Separated'].str.replace(r'[^\w\s]+', "  ")
df3['Punctuation'] = df3['Punctuation'].apply(remove_accented_chars_func)

# III. Turn numbers into text

df3['Convert Numbers to Words'] = df3['Punctuation'].apply(num_to_words)
df3['Convert Numbers to Words'] = df3['Convert Numbers to Words'].str.replace(r'[^\w\s]+', "  ")

# IV. Remove extra white spaces

df3['Convert Numbers to Words'] = df3['Convert Numbers to Words'].apply(remove_extra_whitespaces_func)

# V. Turn all words into lowercase

df3['Lowercase'] = df3['Convert Numbers to Words'].str.lower()

# VI. Remove stop words

df3['Tokenized'] = df3['Lowercase'].apply(wordpunct_tokenize)
df3['Stopwords'] = df3['Tokenized'].apply(lambda x: [word for word in x if word not in stop_words])

# VII. Lemmatize the reviews

df3['POS Tags'] = df3['Stopwords'].apply(nltk.tag.pos_tag)
df3['Wordnet'] = df3['POS Tags'].apply(lambda x: [(word, get_wordnet_pos(pos_tag)) for (word, pos_tag) in x])
df3['Final Pre-processed Review'] = df3['Wordnet'].apply(lambda x: [wnl.lemmatize(word, tag) for word, tag in x])
df3['Final Pre-processed Review'] = df3['Final Pre-processed Review'].apply(lambda x: [word for word in x if word not in stop_words])

# VIII. Output 15 sample pre-processed reviews

df3.drop('Reviewer Name', inplace=True, axis=1)
df3.drop('Rating', inplace=True, axis=1)
dict2 = {'Review': 'Original Review'}
df3.rename(columns=dict2, inplace=True)
df3.drop('Contractions', inplace=True, axis=1)
df3.drop('Back to String', inplace=True, axis=1)
df3.drop('Separated', inplace=True, axis=1)
df3.drop('Punctuation', inplace=True, axis=1)
df3.drop('Convert Numbers to Words', inplace=True, axis=1)
df3.drop('Lowercase', inplace=True, axis=1)
df3.drop('Tokenized', inplace=True, axis=1)
df3.drop('Stopwords', inplace=True, axis=1)
df3.drop('POS Tags', inplace=True, axis=1)
df3.drop('Wordnet', inplace=True, axis=1)
df3.head(15)

<ipython-input-10-5d4e01206090>:12: FutureWarning: The default value of regex will change from True to False in a future version.
  df3['Punctuation'] = df3['Separated'].str.replace(r'[^\w\s]+', "  ")
<ipython-input-10-5d4e01206090>:18: FutureWarning: The default value of regex will change from True to False in a future version.
  df3['Convert Numbers to Words'] = df3['Convert Numbers to Words'].str.replace(r'[^\w\s]+', "  ")


Package Name  \
0   com.peoplefun.wordcross   
1   com.peoplefun.wordcross   
2   com.peoplefun.wordcross   
3   com.peoplefun.wordcross   
4   com.peoplefun.wordcross   
5   com.peoplefun.wordcross   
6   com.peoplefun.wordcross   
7   com.peoplefun.wordcross   
8   com.peoplefun.wordcross   
9   com.peoplefun.wordcross   
10  com.peoplefun.wordcross   
11  com.peoplefun.wordcross   
12  com.peoplefun.wordcross   
13  com.peoplefun.wordcross   
14  com.peoplefun.wordcross   

                                      Original Review  \
0   Gets your mind working putting together puzzle...   
1                          Soothing, challenging, fun   
2                          Very fun and relaxing game   
3                                               Great   
4   Game seems incapable of accurately tracking re...   
5   Fun game, game is a great time killer and make...   
6                                              ggoodd   
7   Good I like it a lot it's just so many interru...   
8                                                  Ok   
9                                  Good and addictive   
10  The awarding of the daily challenge stars gets...   
11  Liar I compled level 38 and no cash out don't ...   
12                          It is hard but also fun👌💯   
13  My mom let me play this game all the time and ...   
14  Daily challenge stars given are often less tha...   

                           Final Pre-processed Review  
0   [get, mind, work, put, together, puzzle, get, ...  
1                            [soothe, challenge, fun]  
2                                        [fun, relax]  
3                                             [great]  
4   [seem, incapable, accurately, track, reward, a...  
5   [fun, great, time, killer, make, think, would,...  
6                                            [ggoodd]  
7               [good, like, lot, many, interruption]  
8                                                [ok]  
9                                   [good, addictive]  
10  [award, daily, challenge, star, get, interrupt...  
11  [liar, compled, level, thirty, eight, cash, do...  
12                                  [hard, also, fun]  
13        [mom, let, play, time, love, also, profile]  
14  [daily, challenge, star, give, often, less, ac...

##### Competitor App 1 - Party Games for Groups

In [ ]:
df4 = pd.read_csv('dh3games.partygamesforgroups.csv')

# I. Remove punctuations
# II. Remove special characters and emojis

df4['Contractions'] = df4['Review'].apply(lambda x: [contractions.fix(word) for word in x.split()])
df4['Back to String'] = [' '.join(map(str, l)) for l in df4['Contractions']]
df4['Back to String'].replace('1st', 'first', regex=True, inplace=True)
df4['Back to String'].replace('2nd', 'second', regex=True, inplace=True)
df4['Back to String'].replace('3rd', 'third', regex=True, inplace=True)
df4['Separated'] = df4['Back to String'].apply(sep_num_words_func)
df4['Punctuation'] = df4['Separated'].str.replace(r'[^\w\s]+', "  ")
df4['Punctuation'] = df4['Punctuation'].apply(remove_accented_chars_func)

# III. Turn numbers into text

df4['Convert Numbers to Words'] = df4['Punctuation'].apply(num_to_words)
df4['Convert Numbers to Words'] = df4['Convert Numbers to Words'].str.replace(r'[^\w\s]+', "  ")

# IV. Remove extra white spaces

df4['Convert Numbers to Words'] = df4['Convert Numbers to Words'].apply(remove_extra_whitespaces_func)

# V. Turn all words into lowercase

df4['Lowercase'] = df4['Convert Numbers to Words'].str.lower()

# VI. Remove stop words

df4['Tokenized'] = df4['Lowercase'].apply(wordpunct_tokenize)
df4['Stopwords'] = df4['Tokenized'].apply(lambda x: [word for word in x if word not in stop_words])

# VII. Lemmatize the reviews

df4['POS Tags'] = df4['Stopwords'].apply(nltk.tag.pos_tag)
df4['Wordnet'] = df4['POS Tags'].apply(lambda x: [(word, get_wordnet_pos(pos_tag)) for (word, pos_tag) in x])
df4['Final Pre-processed Review'] = df4['Wordnet'].apply(lambda x: [wnl.lemmatize(word, tag) for word, tag in x])
df4['Final Pre-processed Review'] = df4['Final Pre-processed Review'].apply(lambda x: [word for word in x if word not in stop_words])

# VIII. Output 15 sample pre-processed reviews

df4.drop('Reviewer Name', inplace=True, axis=1)
df4.drop('Rating', inplace=True, axis=1)
dict2 = {'Review': 'Original Review'}
df4.rename(columns=dict2, inplace=True)
df4.drop('Contractions', inplace=True, axis=1)
df4.drop('Back to String', inplace=True, axis=1)
df4.drop('Separated', inplace=True, axis=1)
df4.drop('Punctuation', inplace=True, axis=1)
df4.drop('Convert Numbers to Words', inplace=True, axis=1)
df4.drop('Lowercase', inplace=True, axis=1)
df4.drop('Tokenized', inplace=True, axis=1)
df4.drop('Stopwords', inplace=True, axis=1)
df4.drop('POS Tags', inplace=True, axis=1)
df4.drop('Wordnet', inplace=True, axis=1)
df4.head(15)

<ipython-input-11-167ee11cfd80>:12: FutureWarning: The default value of regex will change from True to False in a future version.
  df4['Punctuation'] = df4['Separated'].str.replace(r'[^\w\s]+', "  ")
<ipython-input-11-167ee11cfd80>:18: FutureWarning: The default value of regex will change from True to False in a future version.
  df4['Convert Numbers to Words'] = df4['Convert Numbers to Words'].str.replace(r'[^\w\s]+', "  ")


Package Name       Original Review  \
0  dh3games.partygamesforgroups                Hehehe   
1  dh3games.partygamesforgroups  too many advertising   

  Final Pre-processed Review  
0                   [hehehe]  
1        [many, advertising]

##### Competitor App 2 - Quinb

In [ ]:
df5 = pd.read_csv('xyz.deepdaikon.quinb.csv')

# I. Remove punctuations
# II. Remove special characters and emojis

df5['Contractions'] = df5['Review'].apply(lambda x: [contractions.fix(word) for word in x.split()])
df5['Back to String'] = [' '.join(map(str, l)) for l in df5['Contractions']]
df5['Back to String'].replace('1st', 'first', regex=True, inplace=True)
df5['Back to String'].replace('2nd', 'second', regex=True, inplace=True)
df5['Back to String'].replace('3rd', 'third', regex=True, inplace=True)
df5['Separated'] = df5['Back to String'].apply(sep_num_words_func)
df5['Punctuation'] = df5['Separated'].str.replace(r'[^\w\s]+', "  ")
df5['Punctuation'] = df5['Punctuation'].apply(remove_accented_chars_func)

# III. Turn numbers into text

df5['Convert Numbers to Words'] = df5['Punctuation'].apply(num_to_words)
df5['Convert Numbers to Words'] = df5['Convert Numbers to Words'].str.replace(r'[^\w\s]+', "  ")

# IV. Remove extra white spaces

df5['Convert Numbers to Words'] = df5['Convert Numbers to Words'].apply(remove_extra_whitespaces_func)

# V. Turn all words into lowercase

df5['Lowercase'] = df5['Convert Numbers to Words'].str.lower()

# VI. Remove stop words

df5['Tokenized'] = df5['Lowercase'].apply(wordpunct_tokenize)
df5['Stopwords'] = df5['Tokenized'].apply(lambda x: [word for word in x if word not in stop_words])

# VII. Lemmatize the reviews

df5['POS Tags'] = df5['Stopwords'].apply(nltk.tag.pos_tag)
df5['Wordnet'] = df5['POS Tags'].apply(lambda x: [(word, get_wordnet_pos(pos_tag)) for (word, pos_tag) in x])
df5['Final Pre-processed Review'] = df5['Wordnet'].apply(lambda x: [wnl.lemmatize(word, tag) for word, tag in x])
df5['Final Pre-processed Review'] = df5['Final Pre-processed Review'].apply(lambda x: [word for word in x if word not in stop_words])

# VIII. Output 15 sample pre-processed reviews

df5.drop('Reviewer Name', inplace=True, axis=1)
df5.drop('Rating', inplace=True, axis=1)
dict2 = {'Review': 'Original Review'}
df5.rename(columns=dict2, inplace=True)
df5.drop('Contractions', inplace=True, axis=1)
df5.drop('Back to String', inplace=True, axis=1)
df5.drop('Separated', inplace=True, axis=1)
df5.drop('Punctuation', inplace=True, axis=1)
df5.drop('Convert Numbers to Words', inplace=True, axis=1)
df5.drop('Lowercase', inplace=True, axis=1)
df5.drop('Tokenized', inplace=True, axis=1)
df5.drop('Stopwords', inplace=True, axis=1)
df5.drop('POS Tags', inplace=True, axis=1)
df5.drop('Wordnet', inplace=True, axis=1)
df5.head(15)

<ipython-input-12-6b1b8fa161de>:12: FutureWarning: The default value of regex will change from True to False in a future version.
  df5['Punctuation'] = df5['Separated'].str.replace(r'[^\w\s]+', "  ")
<ipython-input-12-6b1b8fa161de>:18: FutureWarning: The default value of regex will change from True to False in a future version.
  df5['Convert Numbers to Words'] = df5['Convert Numbers to Words'].str.replace(r'[^\w\s]+', "  ")


Package Name                                    Original Review  \
0  xyz.deepdaikon.quinb  Very well built app. Taught my little niece ho...   

                          Final Pre-processed Review  
0  [well, build, app, taught, little, niece, play...

##### I have included a screenshot of the sample output of Pre-processed reviews:



See [README](https://github.com/josephspagnuolo/App-Reviews?tab=readme-ov-file#sample-pre-processed-reviews-output) file.

#### **SENTIMENT ANALYSIS**

##### Imports and Defined Functions:

In [ ]:
!pip install -qq vaderSentiment
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
from textblob import TextBlob
analyzer = SentimentIntensityAnalyzer()

def getPolarity(text):
   return TextBlob(text).sentiment.polarity

##### My App - Parlera

In [ ]:
df10 = pd.read_csv('com.enjoyingfoss.parlera.csv')
df10['Final'] = df1['Final Pre-processed Review']
df10['Textblob'] = [' '.join(map(str, l)) for l in df10['Final']]
df10['Polarity'] = df10['Textblob'].apply(getPolarity)
df10.drop('Reviewer Name', inplace=True, axis=1)
df10.drop('Rating', inplace=True, axis=1)
df10.drop('Textblob', inplace=True, axis=1)
df10.drop('Final', inplace=True, axis=1)
df10.head(250)

Package Name  \
0  com.enjoyingfoss.parlera   

                                              Review  Polarity  
0  Super fun game similar to Heads Up, but withou...  0.108333

In [ ]:
df11 = pd.read_csv('com.enjoyingfoss.parlera.csv')
df11['Final'] = df1['Final Pre-processed Review']
df11['Textblob'] = [' '.join(map(str, l)) for l in df11['Final']]
df11['Polarity'] = df11['Textblob'].apply(analyzer.polarity_scores)
df11['Polarity'] = df11['Polarity'].apply(lambda score_dict:score_dict['compound'])
df11.drop('Reviewer Name', inplace=True, axis=1)
df11.drop('Rating', inplace=True, axis=1)
df11.drop('Textblob', inplace=True, axis=1)
df11.drop('Final', inplace=True, axis=1)
df11.head(250)

Package Name  \
0  com.enjoyingfoss.parlera   

                                              Review  Polarity  
0  Super fun game similar to Heads Up, but withou...    0.8884

##### Similar App 1 - Heads Up!

In [ ]:
df20 = pd.read_csv('com.wb.headsup.csv')
df20['Final'] = df2['Final Pre-processed Review']
df20['Textblob'] = [' '.join(map(str, l)) for l in df20['Final']]
df20['Polarity'] = df20['Textblob'].apply(getPolarity)
df20.drop('Reviewer Name', inplace=True, axis=1)
df20.drop('Rating', inplace=True, axis=1)
df20.drop('Textblob', inplace=True, axis=1)
df20.drop('Final', inplace=True, axis=1)
df20.head(250)

Package Name                                             Review  \
0    com.wb.headsup         It wouldn't Load and it kept getting stuck   
1    com.wb.headsup  This game is taking my money this very bad. An...   
2    com.wb.headsup                 It has such a hard time opening...   
3    com.wb.headsup  This app sucks, won't open, downloads about 90...   
4    com.wb.headsup                              does not lode hate it   
..              ...                                                ...   
245  com.wb.headsup                                        doesnt work   
246  com.wb.headsup   Ad page that I couldn't get passed. Uninstalled.   
247  com.wb.headsup  I use to enjoy playing it but it no longer loa...   
248  com.wb.headsup  It won't work right I've tired opening the app...   
249  com.wb.headsup                                 Won't let me play.   

     Polarity  
0    0.000000  
1   -0.700000  
2   -0.145833  
3    0.000000  
4   -0.800000  
..        ...  
245  0.000000  
246  0.000000  
247  0.196591  
248  0.095238  
249  0.000000  

[250 rows x 3 columns]

In [ ]:
df21 = pd.read_csv('com.wb.headsup.csv')
df21['Final'] = df2['Final Pre-processed Review']
df21['Textblob'] = [' '.join(map(str, l)) for l in df21['Final']]
df21['Polarity'] = df21['Textblob'].apply(analyzer.polarity_scores)
df21['Polarity'] = df21['Polarity'].apply(lambda score_dict:score_dict['compound'])
df21.drop('Reviewer Name', inplace=True, axis=1)
df21.drop('Rating', inplace=True, axis=1)
df21.drop('Textblob', inplace=True, axis=1)
df21.drop('Final', inplace=True, axis=1)
df21.head(250)

Package Name                                             Review  \
0    com.wb.headsup         It wouldn't Load and it kept getting stuck   
1    com.wb.headsup  This game is taking my money this very bad. An...   
2    com.wb.headsup                 It has such a hard time opening...   
3    com.wb.headsup  This app sucks, won't open, downloads about 90...   
4    com.wb.headsup                              does not lode hate it   
..              ...                                                ...   
245  com.wb.headsup                                        doesnt work   
246  com.wb.headsup   Ad page that I couldn't get passed. Uninstalled.   
247  com.wb.headsup  I use to enjoy playing it but it no longer loa...   
248  com.wb.headsup  It won't work right I've tired opening the app...   
249  com.wb.headsup                                 Won't let me play.   

     Polarity  
0     -0.2500  
1     -0.7717  
2     -0.1027  
3     -0.7299  
4     -0.5719  
..        ...  
245    0.0000  
246    0.0000  
247    0.7906  
248    0.0000  
249    0.3400  

[250 rows x 3 columns]

##### Similar App 2 - Wordscapes

In [ ]:
df30 = pd.read_csv('com.peoplefun.wordcross.csv')
df30['Final'] = df3['Final Pre-processed Review']
df30['Textblob'] = [' '.join(map(str, l)) for l in df30['Final']]
df30['Polarity'] = df30['Textblob'].apply(getPolarity)
df30.drop('Reviewer Name', inplace=True, axis=1)
df30.drop('Rating', inplace=True, axis=1)
df30.drop('Textblob', inplace=True, axis=1)
df30.drop('Final', inplace=True, axis=1)
df30.head(250)

Package Name  \
0    com.peoplefun.wordcross   
1    com.peoplefun.wordcross   
2    com.peoplefun.wordcross   
3    com.peoplefun.wordcross   
4    com.peoplefun.wordcross   
..                       ...   
245  com.peoplefun.wordcross   
246  com.peoplefun.wordcross   
247  com.peoplefun.wordcross   
248  com.peoplefun.wordcross   
249  com.peoplefun.wordcross   

                                                Review  Polarity  
0    Gets your mind working putting together puzzle... -0.195833  
1                           Soothing, challenging, fun  0.300000  
2                           Very fun and relaxing game  0.300000  
3                                                Great  0.800000  
4    Game seems incapable of accurately tracking re...  0.054167  
..                                                 ...       ...  
245  Why aren't my stars counting anymore? I'll get...  0.250000  
246  Just uninstalled the game after getting to lev... -0.010011  
247  Don't force us to use gems on animals just to ...  0.000000  
248                            Gets better and better!  0.500000  
249                          So enjoyable to relax too  0.500000  

[250 rows x 3 columns]

In [ ]:
df31 = pd.read_csv('com.peoplefun.wordcross.csv')
df31['Final'] = df3['Final Pre-processed Review']
df31['Textblob'] = [' '.join(map(str, l)) for l in df31['Final']]
df31['Polarity'] = df31['Textblob'].apply(analyzer.polarity_scores)
df31['Polarity'] = df31['Polarity'].apply(lambda score_dict:score_dict['compound'])
df31.drop('Reviewer Name', inplace=True, axis=1)
df31.drop('Rating', inplace=True, axis=1)
df31.drop('Textblob', inplace=True, axis=1)
df31.drop('Final', inplace=True, axis=1)
df31.head(250)

Package Name  \
0    com.peoplefun.wordcross   
1    com.peoplefun.wordcross   
2    com.peoplefun.wordcross   
3    com.peoplefun.wordcross   
4    com.peoplefun.wordcross   
..                       ...   
245  com.peoplefun.wordcross   
246  com.peoplefun.wordcross   
247  com.peoplefun.wordcross   
248  com.peoplefun.wordcross   
249  com.peoplefun.wordcross   

                                                Review  Polarity  
0    Gets your mind working putting together puzzle...   -0.1027  
1                           Soothing, challenging, fun    0.7269  
2                           Very fun and relaxing game    0.7351  
3                                                Great    0.6249  
4    Game seems incapable of accurately tracking re...    0.1779  
..                                                 ...       ...  
245  Why aren't my stars counting anymore? I'll get...    0.3612  
246  Just uninstalled the game after getting to lev...    0.5574  
247  Don't force us to use gems on animals just to ...    0.3400  
248                            Gets better and better!    0.6124  
249                          So enjoyable to relax too    0.7003  

[250 rows x 3 columns]

##### Competitor App 1 - Party Games for Groups

In [ ]:
df40 = pd.read_csv('dh3games.partygamesforgroups.csv')
df40['Final'] = df4['Final Pre-processed Review']
df40['Textblob'] = [' '.join(map(str, l)) for l in df40['Final']]
df40['Polarity'] = df40['Textblob'].apply(getPolarity)
df40.drop('Reviewer Name', inplace=True, axis=1)
df40.drop('Rating', inplace=True, axis=1)
df40.drop('Textblob', inplace=True, axis=1)
df40.drop('Final', inplace=True, axis=1)
df40.head(250)

Package Name                Review  Polarity
0  dh3games.partygamesforgroups                Hehehe       0.0
1  dh3games.partygamesforgroups  too many advertising       0.5

In [ ]:
df41 = pd.read_csv('dh3games.partygamesforgroups.csv')
df41['Final'] = df4['Final Pre-processed Review']
df41['Textblob'] = [' '.join(map(str, l)) for l in df41['Final']]
df41['Polarity'] = df41['Textblob'].apply(analyzer.polarity_scores)
df41['Polarity'] = df41['Polarity'].apply(lambda score_dict:score_dict['compound'])
df41.drop('Reviewer Name', inplace=True, axis=1)
df41.drop('Rating', inplace=True, axis=1)
df41.drop('Textblob', inplace=True, axis=1)
df41.drop('Final', inplace=True, axis=1)
df41.head(250)

Package Name                Review  Polarity
0  dh3games.partygamesforgroups                Hehehe       0.0
1  dh3games.partygamesforgroups  too many advertising       0.0

##### Competitor App 2 - Quinb

In [ ]:
df50 = pd.read_csv('xyz.deepdaikon.quinb.csv')
df50['Final'] = df5['Final Pre-processed Review']
df50['Textblob'] = [' '.join(map(str, l)) for l in df50['Final']]
df50['Polarity'] = df50['Textblob'].apply(getPolarity)
df50.drop('Reviewer Name', inplace=True, axis=1)
df50.drop('Rating', inplace=True, axis=1)
df50.drop('Textblob', inplace=True, axis=1)
df50.drop('Final', inplace=True, axis=1)
df50.head(250)

Package Name                                             Review  \
0  xyz.deepdaikon.quinb  Very well built app. Taught my little niece ho...   

   Polarity  
0  0.328125

In [ ]:
df51 = pd.read_csv('xyz.deepdaikon.quinb.csv')
df51['Final'] = df5['Final Pre-processed Review']
df51['Textblob'] = [' '.join(map(str, l)) for l in df51['Final']]
df51['Polarity'] = df51['Textblob'].apply(analyzer.polarity_scores)
df51['Polarity'] = df51['Polarity'].apply(lambda score_dict:score_dict['compound'])
df51.drop('Reviewer Name', inplace=True, axis=1)
df51.drop('Rating', inplace=True, axis=1)
df51.drop('Textblob', inplace=True, axis=1)
df51.drop('Final', inplace=True, axis=1)
df51.head(250)

Package Name                                             Review  \
0  xyz.deepdaikon.quinb  Very well built app. Taught my little niece ho...   

   Polarity  
0    0.8465

##### PART V:

How does the sentiments retrieved by Textblob and Vader compare with each
other? How do you interpret the similarity/difference? Which one is the best option
for review analysis of your apps? Why?

Sample Textblob Output:

See [README](https://github.com/josephspagnuolo/App-Reviews?tab=readme-ov-file#sample-textblob-sentiment-analysis-output) file.

Sample Vader output:

See [README](https://github.com/josephspagnuolo/App-Reviews?tab=readme-ov-file#sample-vader-sentiment-analysis-output) file.

We will be comparing the Sentiment Analysis from Textblob and Vader, discussing their similarities and differences and their overall effective for app reviews. When looking at the polartity number on a scale of -1 to 1, we say that -1 is very negative, 1 is very positive, and 0 is neutral. For some reviews, Textblob and Vader give similar polarities, such as review 2 which outputted -0.145833 and -0.1027, respectively. This means that this review was somewhat negative. Similarly, for review 245 and 246, they both outputted 0.0000 meaning the review is neutral.



However, on the other hand, looking at review 3 gives a different perspective. We can see that the review explicitly says "this app sucks", and so we would naturally be under the impression that this is a negative review. The output from Vader agrees with this, giving a -0.7299 polarity score, but confusingly, Textblob gives a 0.0000. This is clearly an uneffective representation of the review polarity, resulting in inaccurate feedback analysis. It is because of this that I would say Vader is more effective.

#### **TOPIC MODELLING USING LDA**

##### Heads Up!

In [ ]:
from collections import defaultdict
import gensim
from gensim import corpora
from gensim.corpora.dictionary import Dictionary

dictionary = Dictionary(df2['Final Pre-processed Review'])
dictionary.filter_extremes(no_below=5, no_above=0.5, keep_n=1000)
corpus = [dictionary.doc2bow(doc) for doc in df2['Final Pre-processed Review']]
Lda = gensim.models.ldamodel.LdaModel
ldamodel = Lda(corpus, num_topics=10, id2word = dictionary, passes=10, random_state=0, eval_every=None)
for idx, topic in ldamodel.show_topics(formatted=False, num_words= 10):
    print('Topic {} described by Words: {}\n'.format(idx+1, ', '.join([w[0] for w in topic])))

Topic 1 described by Words: play, load, take, get, cannot, even, back, long, let, screen

Topic 2 described by Words: play, family, fun, great, friend, love, best, ever, time, good

Topic 3 described by Words: love, star, give, would, could, one, zero, kid, ellen, wish

Topic 4 described by Words: crash, open, app, work, good, even, keep, time, suck, four

Topic 5 described by Words: phone, app, work, download, time, open, try, go, play, cool

Topic 6 described by Words: need, hate, ad, get, app, freeze, well, download, fix, review

Topic 7 described by Words: one, two, think, hundred, add, eight, party, five, pas, l

Topic 8 described by Words: fun, really, like, much, good, lot, play, super, thing, say

Topic 9 described by Words: deck, money, pay, buy, purchase, pack, free, get, one, like

Topic 10 described by Words: load, bad, app, never, nice, work, even, would, please, fix



##### PART II:

Screenshot of the LDA topic modelling output with chosen topic names:

See [README](https://github.com/josephspagnuolo/App-Reviews?tab=readme-ov-file#sample-lda-topic-modelling-output) file.

##### PART III:
How do you explain the differences in terms of accuracy of the model and the size
of data? 

'Heads Up!' is a very popular app with over 10M+ downloads and so the size of the data is very large. We have looked at 5000 reviews from the Google Play Store. Because we have so much data, statistically it should be true that the analyzed reviews are a representative sample of the user base. However, these reviews are very short in length, and often do not give specific information about software functionality. So simply put, the accuracy of the model may be limited by the lack of detailed reviews. We can see that many user reviews from the Google Play store are basic expressions of frustration about the app crashing, or similarly expressions of enjoyment from playing in general. This means that many reviews will not go into detail about all features, and so there may be some variance from the feature clusters we have identified earlier.

##### PART IV:
Does any of the topics relate to the app's core functionalities? Why? 


Topics 2, 3, 5, 7, and 8 more directly relate to the core functionality of our SUD. They relate because the app's goal is to be a fun game to play among groups of people. This is shown in the app description and screenshots, and so many users make feedback on this. However, there is certainly a lack of detail regarding the core functionalities. Many users do not mention the feature of tilting up or down to signal right or wrong, even though this is a core feature of the application. They may not give feedback on specific features because they are focused on other things like expensive in-app purchases for example, or they are only looking to leave a general review rather than a comprehensive one.